cmd/chartsvc/handler.go

```diff
@@ -444,3 +445,75 @@ func newChartVersionListResponse(c *models.Chart) apiListResponse {
 
 	return cvl
 }
+
+func handleRedirect(w http.ResponseWriter, req *http.Request) {
+
+	// The URL can be in one of two forms:
+	// - /v1/redirect/charts/stable/aerospike
+	// - /v1/redirect/charts/stable/aerospike/v1.2.3
+	// And either of these can optionally have a trailing /
+
+	// Make sure the path is valid
+	ct := strings.TrimPrefix(req.URL.Path, "/v1/redirect/charts/")
+
+	// check if URL for provenance
+	prov := strings.HasSuffix(ct, ".prov")
+	ct = strings.TrimSuffix(ct, ".prov")
+
+	ct = strings.TrimSuffix(ct, "/") // Removing the optional / on the end
+	p := strings.Split(ct, "/")
+
+	// Not enough parts passed in to the path
+	if len(p) < 2 || len(p) > 3 {
+		response.NewErrorResponse(http.StatusNotFound, "could not find chart").Write(w)
+		return
+	}
+
+	// Decide if latest or a version
+	var version string
+	if len(p) == 3 {
+		version = p[2]
+	}
+
+	// Look it up. This will be different if there is a version or we are getting
+	// the latest
+	db, closer := dbSession.DB()
+	defer closer()
+	var chart models.Chart
+	chartID := fmt.Sprintf("%s/%s", p[0], p[1])
+
+	if version == "" {
+		if err := db.C(chartCollection).FindId(chartID).One(&chart); err != nil {
```

Comment:
Since we're starting to repeat these MongoDB calls, we may want to extract them out to separate functions and call them from here and in `getChart` (and the one below in `getChartVersion`)

Reply:
Can we do this as part of a separate change? This query is used 4 times by 4 different callbacks. If the logic were to be extracted I would suggest doing it uniformly and updating all of them. It's worth being tested separately from this feature addition IMHO. Repeated the same database query is common here as it stands.